### PR TITLE
Delete tenant's data from s3 

### LINKED ISSUE #4250: Implement deletion as described in RFC for tenants

### PR DESCRIPTION
**Pull request by LizardWizzard**
## Summary of changes

For context see https://github.com/neondatabase/neon/blob/main/docs/rfcs/022-pageserver-delete-from-s3.md

Create Flow to delete tenant's data from pageserver. The approach heavily mimics previously implemented timeline deletion implemented mostly in https://github.com/neondatabase/neon/pull/4384 and followed up in https://github.com/neondatabase/neon/pull/4552 

For remaining deletion related issues consult with deletion project here: https://github.com/orgs/neondatabase/projects/33

resolves #4250
resolves https://github.com/neondatabase/neon/issues/3889

### CODE REVIEW

File: pageserver/src/http/routes.rs
Comment (arpad-m) [inline pageserver/src/http/routes.rs:641]:
```suggestion
        .instrument(info_span!("tenant_delete_handler", tenant = %tenant_id))
```

**Review by koivunej — COMMENTED** [review]
First round of questions, suggestions

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:327]: This only prevents deletion from interfering with startup. We still allow to delete broken tenants.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:854]:
```suggestion
                if let Some(deletion) = pending_deletion {
                    // as we are no longer loading, signal completion by dropping
                    // the completion while we resume deletion
                    drop(_completion);
                    match DeleteTenantFlow::resume(
```

Otherwise it should hang on startup, because init_done_rx never completes in `pageserver/src/main.rs`, if the `background_jobs_can_start` are used within this path.

Comment (koivunej) [reply pageserver/src/tenant.rs:854]: the barriers are an obvious deadlock hazard... should create a crate which would panic if you wait on a barrier while keeping up a guard. Problem is of course how to do this transitively...

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:298]: Deletion has the lowest priority among startup tasks, so delaying it as much as possible to leave room for more important stuff makes sense to me

File: pageserver/src/tenant/timeline/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline/delete.rs]: Hm, thats a good question. Even ephemeral files shouldnt be an issue. They're deleted when all references are dropped and here we still hold timeline Arc. So deletion of ephemeral files will happen after deletion flow completion.

Probably I should just remove that code and leave general error for all cases

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:855]: Good catch!

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs]: Done https://github.com/neondatabase/neon/pull/4855/commits/5a42259332353cb6506fa0796bc90548613a80a8

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs]: Done https://github.com/neondatabase/neon/pull/4855/commits/5a42259332353cb6506fa0796bc90548613a80a8

**Comment by LizardWizzard** [comment]
Thanks for review! I resolved some of the comments to more clearly see remaining ones, feel free to unresolve if needed

File: pageserver/src/tenant/timeline/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline/delete.rs]: Removed the code

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:423]: Ok, thank you! Still using protection inside shutdown simultaneously with deletion guard may look a bit confusing. See also https://github.com/neondatabase/neon/pull/4855/files#diff-100b728a9529fbc86c8c054a1d7d3d0b5bc129a7ce27d28d833f60b7817169f7R338

**Comment by LizardWizzard** [comment]
Existing tests seem to pass, continue with adding more deletion specific ones

**Comment by LizardWizzard** [comment]
Fixed one more `real_s3` flakiness. I think this is ready for review. I have in mind two more test cases, they're rather small, and can go in here or in a follow up.

File: pageserver/src/tenant/delete.rs
Comment (jcsp) [inline pageserver/src/tenant/delete.rs:137]: I would usually expect an "assert" function to be something that panics if condition isn't met -- it might be more obvious to make this `is_timelines_dir_empty` and return `Result<bool, io::Error>` -- that way we can distinguish successful I/O where the directory is non-empty from actual I/O errors.

File: pageserver/src/http/openapi_spec.yml
Comment (jcsp) [inline pageserver/src/http/openapi_spec.yml:99]: Hmm.  404 could also mean that someone detached the tenant, right?

I wonder if we should just leave remote deletion markers in place forever (they're cheap) and use that to be able to unambiguously distinguish between a deletion being complete and a tenant just not being attached for some other reason.  It would also be useful if there is some bug in deletion, and we later want to clean up some layer files -- it enables us to be certain that a particular tenant was intentionally deleted.

**Comment by jcsp** [comment]
It's not changed in this PR, but reading through the deletion path I noticed that most of the deletions are happening with one DeleteObject requests per object, plural DeleteObjects is only used for the `remaining` part of `delete_all`.

That also makes me wonder if we can avoid some complexity and runtime cost by simply doing the `list_prefixes` & deletion part of `delete_all` -- just avoid doing any timeline-aware/metadata-aware stuff, and wipe out all the files?  Since if we see any remaining files we'll nuke them anyway in the existing latter part of `RemoteTimelineClient::delete_all`

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:141]: Note that this is not the directory where all layers live. Its `<tenant>/timelines` directory. So there shouldnt be many entries. If no objections I'd keep it shorter/simpler.

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:137]: I'll change the name to start with `ensure` instead of assert.

File: pageserver/src/http/openapi_spec.yml
Comment (arpad-m) [reply pageserver/src/http/openapi_spec.yml:99]: IIRC the scheme this PR is proposing is already used by timeline deletions (even though it's not documented in the very same way as tenant deletions).

Comment (LizardWizzard) [reply pageserver/src/http/openapi_spec.yml:99]: > Hmm. 404 could also mean that someone detached the tenant, right?

That sounds correct. Though I wont expect that. If tenant is being deleted why detach it in the first place? It wont be used ever again. Detach calls to `tenant.shutdown` which will error out because tenant is already in stopping state. In all remaining cases 404 indeed means successful deletion. 

> I wonder if we should just leave remote deletion markers in place forever (they're cheap)

This was brought up in RFC discussion I think, but IMO we should try to maintain zero garbage policy.
1. It still costs something.
2. It makes investigations trickier (you need to dig through the pile of garbage to get what you need)
3. ListObjects becomes slower and slower if number of entries keeps increasing. We dont do list on `tenants/` during normal operations, but it is useful in manual exploration. This can appear in case we add some background gc-like operation that walks every tenant for stuff to delete like you propose in your RFC

**Comment by LizardWizzard** [comment]
> but reading through the deletion path I noticed that most of the deletions are happening with one DeleteObject requests per object, plural DeleteObjects is only used for the remaining part of delete_all

Right. The fix is tracked here https://github.com/neondatabase/neon/issues/4325.

> That also makes me wonder if we can avoid some complexity and runtime cost by simply doing the list_prefixes & deletion part of delete_all -- just avoid doing any timeline-aware/metadata-aware stuff, and wipe out all the files? Since if we see any remaining files we'll nuke them anyway in the existing latter part of RemoteTimelineClient::delete_all

ListObjects can take quite some time (many s3 roundtrips depending on the number of layers). So I'm not sure why it should be cheaper in terms of runtime cost. Additionally we already have all layers listed in RemoteTimelineClient internals, so to me it doesnt look like there is a big difference where to get the list from.

> we'll nuke them anyway in the existing latter part

Here there shouldnt be many files so ListObjects shouldnt take long. Also IMO we should aim for this call to be noop. I e avoid leaked files at all.

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs]: https://github.com/neondatabase/neon/pull/4855/commits/af2e26f0f4f1d630242e00e680934bf0ebf60ffc

File: pageserver/src/tenant/delete.rs
Comment (jcsp) [reply pageserver/src/tenant/delete.rs:142]: Fair enough -- I'd still like to get rid of the non-streaming list_dir funcntion in future to avoid risk of its misuse, but you're right that it's fine here.

**Comment by jcsp** [comment]
>ListObjects can take quite some time (many s3 roundtrips depending on the number of layers). So I'm not sure why it should be cheaper in terms of runtime cost. Additionally we already have all layers listed in RemoteTimelineClient internals, so to me it doesnt look like there is a big difference where to get the list from.

I guess the main difference would be simplicity: one could write a very short function that just streams the objects from a ListObjectsv2 call into something that buffers them up into DeleteObjects calls.  It's true that there's some overhead to using the listing instead of just using the list of objects we already hold in memory.  Using the list of layers we have in memory is cheaper if the layer count is large, I suppose, whereas for a timeline with few layers, it would be cheaper to do everything in one DeleteObjects rather than having the separate delete, cleanup steps.

But yeah, it's kind of debatable either way, and anyway not directly touched in this PR.

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs]: Agreed, resolving

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs]: Do you have any recollection of where the see also could had pointed (the link is probably outdated).

File: pageserver/src/tenant/timeline/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline/delete.rs]: Ok.

File: pageserver/src/tenant/timeline/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline/delete.rs]: Good

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs:142]: Thinking about this since we are constantly pushing to have bigger tenants than what we really support. 1TB will be 3-5k layers.

Looking at the current tokio impl it's hosting entries in a VecDeque, and `spawn_blocking` to fill to capacity... It might be that we are getting a huge number of entries compared to how many we display.

I'll just reopen this.

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:1812]: Could you include in which cases we (do not)? want to allow the transition from loading? I haven't yet looked at the callers.

**Comment by LizardWizzard** [comment]
>  it would be cheaper to do everything in one DeleteObjects rather than having the separate delete, cleanup steps.

Its a good idea! I havent thought about that. But yeah, unfortunately this doesnt scale to bigger number of layers, so we'd need to maintain two code paths in that case. We can consider taking this shortcut in the future. Also for that to work we need to be sure that DeleteObjects is as atomic as single DeleteObject. S3 [Consistency model](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html#ConsistencyModel) says:

> Amazon S3 provides strong read-after-write consistency for PUT and DELETE requests of objects in your Amazon S3 bucket in all AWS Regions.

And we should keep in mind that we shouldnt rely on AWS-only features of s3 API to make it easier to adapt to s3 API implementations in other cloud providers

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:1812]: In case of resumed deletion we need allow the transition from Loading to Stopping. Previously it was not allowed. So set_stopping was waiting for loading to complete and thus blocked forever

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs]: Hmm, yeah, the link is broken, points to nowhere. I think it was pointing to `Tenant::shutdown` call in `DeleteTenantFlow` here: https://github.com/zenithdb/zenith/blob/af2e26f0f4f1d630242e00e680934bf0ebf60ffc/pageserver/src/tenant/delete.rs#L437

File: test_runner/regress/test_timeline_delete.py
Comment (koivunej) [inline test_runner/regress/test_timeline_delete.py:288]: Thanks!

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [inline pageserver/src/tenant/delete.rs:521]: I don't understand why we fail here? Would/could this cause problems later at `cleanup_remaining_fs_traces` or ... Why?

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [inline pageserver/src/tenant/mgr.rs:85]: Is this still valid? Could expand on what is missing?

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs:521]: This is rather a sanity check. We shouldnt fail here. This will help in case our fs layout changes, maybe new files are introduced etc

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [inline pageserver/src/tenant/delete.rs:375]: Could this be explained somehow so that it wouldn't look like it's just ... quick way to call `set_stopping`? (We may have already covered this but) I see following possibilities to enhance this:
- we keep hold of this `_` until the background operation completes
  - if we force it to cover the spawned delete upload ops, we might get completly rid of the shutdown problems (which I still haven't understood how did you solve them)
- you could use `Default::default()` as the argument (or `completion::Barrier::default()`) which would be more explicit that this is not a forgotten (thus requiring less explanation)

Comment (koivunej) [reply pageserver/src/tenant/delete.rs:375]: Ok the other instance has comment above it, just now noticed it.

What was the rationale for not holding up shutdown?

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs:521]:
```suggestion
            // sanity check to guard against layout changes
            ensure_timelines_dir_empty(&timelines_path)
                .await
                .context("timelines dir not empty")?;
```

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1812]: Aah good point!

File: pageserver/src/tenant/mgr.rs
Comment (koivunej) [reply pageserver/src/tenant/mgr.rs]: [e2a0938](https://github.com/neondatabase/neon/pull/4855/commits/e2a0938ae664271d4d86e2dcc04458b824a1bcc8) resolving

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs:375]: Okay, with this rationale it makes sense to me. It should only ever matter to the really near shutdown deleted. And with this rationale we should just have the `allowed_errors` for the deletion tasks in some test config which I'm yet to come across.

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:1812]:
```suggestion
    ///
    /// `allow_transition_from_loading` is needed for the special case of loading task deleting the tenant.
    async fn set_stopping(
```

File: pageserver/src/tenant.rs
Comment (koivunej) [inline pageserver/src/tenant.rs:2276]: Was there a discussion you wanted to have about this? Unsure why we need to return default here, could just go with None which would seem to be more honest.

Wouldn't it move the handling to somewhere with more context? Or perhaps not. I think we currently create a tenant config for all tenants... And I think we can also receive the config via attach, so the above FIXME is a bit outdated anyways.

> OR: we're loading after incomplete deletion that managed to remove config.

Is this possible, mark file is removed *after* config?

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs:142]: 5-10k strings ... Seems unfortunate for this, but this is rare path. And we are not likely to have the 5-10k strings here anymore, this is a special case.

Resolving.

File: test_runner/fixtures/pageserver/utils.py
Comment (koivunej) [inline test_runner/fixtures/pageserver/utils.py:257]: Was this forgotten, did you still want to do this?

**Review by koivunej — APPROVED** [review]
All in all I don't think it's useful to spend *more* time looking at this in review. We should merge it today and then spend time debugging any test flakyness if such arise. The `DELETE`'s will not be issued from control plane anyways soon. Do you know the issue tracking that progress?

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs]: Reproduced:

```rust
// It would be good to only set stopping here and continue shutdown in the background, but shutdown is not idempotent.
// i e it is an error to do:
// tenant.set_stopping
// tenant.shutdown
// Its also bad that we're holding tenants.read here.
// TODO relax set_stopping to be idempotent?
if tenant.shutdown(progress, false).await.is_err() {
    return Err(DeleteTenantError::Other(anyhow::anyhow!(
        "tenant shutdown is already in progress"
    )));
}
```

> TODO: set_stopping to be idempotent

This sounds like a sensible idea to me, and then shutdown would be idempotent as there is no task to await for.

Unless in deletion case there might be the tasks to wait for, wouldn't this change that?

File: pageserver/src/tenant/delete.rs
Comment (koivunej) [reply pageserver/src/tenant/delete.rs]: I am good with that.

File: pageserver/src/tenant/mgr.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/mgr.rs]: Thanks! Pushed commit that removes it but forgot to reply here :)

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:2276]: > Was there a discussion you wanted to have about this?

Sort of. To me this looks weird and indeed returning option would be more honest. From outside perspective when looking at `load_tenant_config` I'd imagine that it wont introduce a default one out of thin air. So I consider this as a hack, but one that appeared before this PR. This is why I decided to bring more attention to it. 

So for now lets leave it as is and it can be turned into option in a follow up. Objections?

> Is this possible, mark file is removed after config?

Yeah, this happens because mark file is removed last, so we can observe arbitrary state prior to mark file deletion

**Comment by LizardWizzard** [comment]
> We should merge it today and then spend time debugging any test flakyness if such arise.

Also I'd take a look at startup times because with this PR we'll do one more s3 request per tenant

>  The DELETE's will not be issued from control plane anyways soon. Do you know the issue tracking that progress?

I plan to look into it after this PR gets merged. I added this to deletion project but didnt create an issue yet: https://github.com/orgs/neondatabase/projects/33/views/1

***

So I do one final cleanup pass to remove fixmes and when CI passes I'll merge the PR. Thanks!

File: test_runner/fixtures/pageserver/utils.py
Comment (LizardWizzard) [reply test_runner/fixtures/pageserver/utils.py:257]: > did you still want to do this?

Ideally yes, will take a look, if the change is small enough I'll do it right there, if not I'll do that separately. Thanks!

File: pageserver/src/tenant/delete.rs
Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs]: Waiting for tasks should be also idempotent. If there are any -> wait for them if not -> continue. 

> Unless in deletion case there might be the tasks to wait for, wouldn't this change that?

If deletion fails I'd expect that it wont leave dangling tasks behind. Though this may not always hold. Since every upload queue op ends up in separate task they can outlive main deletion task. If we look at that from shutdown idempotency standpoint it should be fine to wait for such tasks to complete as well. 

I add this to follow up list. This is not critical at all, and currently nobody is calling delete from CP side

Comment (LizardWizzard) [reply pageserver/src/tenant/delete.rs]: Opened separate issue to track this https://github.com/neondatabase/neon/issues/4951

File: test_runner/fixtures/pageserver/utils.py
Comment (LizardWizzard) [reply test_runner/fixtures/pageserver/utils.py]: See https://github.com/neondatabase/neon/pull/4855/commits/d7f2abf68b8daf05bd3043914a687992080c5174

File: pageserver/src/tenant.rs
Comment (koivunej) [reply pageserver/src/tenant.rs:2278]: Hmm okay to leave as is. `Attention ^' looks a bit weird and inactionable, but ... not as bad as likely outdated FIXME.

Yeah it is not clear what to do here.

> > Was there a discussion you wanted to have about this?
> 
> Sort of. To me this looks weird and indeed returning option would be more honest. From outside perspective when looking at `load_tenant_config` I'd imagine that it wont introduce a default one out of thin air. So I consider this as a hack, but one that appeared before this PR. This is why I decided to bring more attention to it.
> 
> So for now lets leave it as is and it can be turned into option in a follow up. Objections?
> 
> > Is this possible, mark file is removed after config?
> 
> Yeah, this happens because mark file is removed last, so we can observe arbitrary state prior to mark file deletion

I did not actually see this possibility but okay now I understand, we could have a crash in-between. Unsure if we have any chance of fixing this without a log and some atomicity. But unsure if I'll enjoy fixing a custom log either :) But it would be just a one place to fix.

Okay to resolve.

**Review by koivunej — APPROVED** [review]

File: pageserver/src/tenant.rs
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:2278]: Makes sense to remove `Attention ^`, will do.

> Unsure if we have any chance of fixing this without a log and some atomicity

Yeah, mark file is poor'mans log consisting of one entry :) Expect this to be improved under the scope of https://github.com/neondatabase/neon/issues/4636

**Pull request merged**
